vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,615 @@
+        // Note: Instead of `&*` we can use `AtomicU8::from_ptr()` as soon it gets stabilized.
+        // SAFETY: `self.addr` + `index` doesn't wrap around and is contained within the mapped
+        // memory region.
+        unsafe { &*self.addr.add(index) }
+    }
+}
+
+impl Drop for MmapLogReg {
+    fn drop(&mut self) {
+        // SAFETY: `addr` is properly aligned, also we are sure that this is the
+        // last reference alive and/or we have an exclusive access to this object.
+        unsafe {
+            libc::munmap(self.addr as *mut libc::c_void, self.len as libc::size_t);
+        }
+    }
+}
+
+trait IoTryInto<T: TryFrom<Self>>: Sized {
+    fn io_try_into(self) -> io::Result<T>;
+}
+
+impl<TySrc, TyDst> IoTryInto<TyDst> for TySrc
+where
+    TyDst: TryFrom<TySrc>,
+    <TyDst as TryFrom<TySrc>>::Error: Send + Sync + std::error::Error + 'static,
+{
+    fn io_try_into(self) -> io::Result<TyDst> {
+        self.try_into()
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    }
+}
+
+#[inline]
+// Get the page number corresponding to the address `addr`
+fn page_number(addr: usize) -> usize {
+    addr / LOG_PAGE_SIZE
+}
+
+#[inline]
+// Get the block within the bitmap of the page.
+// Each page is indexed inside a block of `LOG_WORD_SIZE` bits.
+fn page_word(page: usize) -> usize {
+    page / LOG_WORD_SIZE
+}
+
+#[inline]
+// Get the bit index inside a block of `LOG_WORD_SIZE` bits
+fn page_bit(page: usize) -> usize {
+    page % LOG_WORD_SIZE
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs::File;
+    use std::io::Write;
+    use std::os::fd::AsFd;
+    use vm_memory::{GuestAddress, GuestRegionMmap};
+    use vmm_sys_util::tempfile::TempFile;
+
+    // Helper method to check whether a specified range is clean.
+    pub fn range_is_clean<B: Bitmap>(b: &B, start: usize, len: usize) -> bool {
+        (start..start + len).all(|offset| !b.dirty_at(offset))
+    }
+
+    // Helper method to check whether a specified range is dirty.
+    pub fn range_is_dirty<B: Bitmap>(b: &B, start: usize, len: usize) -> bool {
+        (start..start + len).all(|offset| b.dirty_at(offset))
+    }
+
+    fn tmp_file(len: usize) -> File {
+        let mut f = TempFile::new().unwrap().into_file();
+        let buf = vec![0; len];
+        f.write_all(buf.as_ref()).unwrap();
+        f
+    }
+
+    fn test_all(b: &BitmapMmapRegion, len: usize) {
+        assert!(range_is_clean(b, 0, len), "The bitmap should be clean");
+
+        b.mark_dirty(0, len);
+        assert!(range_is_dirty(b, 0, len), "The bitmap should be dirty");
+    }
+
+    #[test]
+    #[cfg(not(miri))] // Miri cannot mmap files
+    fn test_bitmap_region_bigger_than_log() {
+        let mmap_offset: u64 = 0;
+        let mmap_size = 1; // 8 pages
+        let f = tmp_file(mmap_size);
+
+        let region_start_addr = GuestAddress(mmap_offset);
+        let region_len = LOG_PAGE_SIZE * LOG_WORD_SIZE * mmap_size * 2; // 16 pages
```

Comment:
1 bit in the bitmap correspond to 1 page tracked, since we use and array of `AtomicU8` each byte/word tracks 8 pages.

I updated the comments and the tests, I hope is clearer now